### ISSUE TITLE
ENH: Formatting Polynomials

### ISSUE THREAD

**Issue by philer** [numpy/numpy#8893]
**TL;DR** Add a way of turning `numpy.polynomial.polynomial.Polynomial` instances into a human readable and/or machine parseable string, such as

* `2x³ + x² - 1` (utf8)
* `2x^3 + x^2 - 1` (ascii)
* `2x^{3} + x^{2} - 1` (latex)
* `2x<sup>3</sup> + x<sup>2</sup> - 1` (html)
* `2*x**3 + x**2 - 1` (python)

---

*Disclaimer: I'm relatively new to numpy so if this is just completely out of the question please bear with me and let me know. If this issue finds approval I'd be happy to implement it myself. :)*

**Motivation**

Numpy is used for scientific work all around. In literature polynomials are typically written in the form *2x³ + x² - 1* which I will refer to as "human readable". It is characterized by descending exponents and omission of unnecessary parts involving coefficients and exponents ±1 and 0. Yet neither `str` nor `format` return a version close to the above, instead displaying the "machine oriented" internal representation as an array of coefficients sorted by ascending index. This representation can be confusing for people who only know the human readable form and in some cases completely impractical (e.g. for sparse polynomials of high degree, like *x¹²³*). It would be very useful for authors who employ numpy to be able to generate output directly in a human readable format or in a common markup language (such as LaTeX and HTML). This allows automation and prevents common slipups when manually translating between the formats.

**Possible approaches**

I can think of several ways to incorporate a formatting option into numpy. Personally I prefer option *1)* as it is the most pythonic and I can not see any significant downsides assuming the default is chosen to avoid backwards compatibility issues at least for upcoming minor versions. Some of these options may also coexist.

*1)* Implement `Polynomial.__format__` with a reasonable default. In this case "reasonable" could mean either defaulting to `str` if no style option is provided (in the interest of backward compatibility), or defaulting to one of the versions in the TL;DR example.

As a result the following code:
```Python
p = Polynomial([-1, 0, 1, 2])
print(format(p))       # default to str(p)
print("{:ascii}".format(p))
print(f"{p :utf8}")    # Python 3.6+
```
… could output:
```
poly([-1.  0.  1.  2.])
2x^3 + x^2 - 1
2x³ + x² - 1
```

*2)* Add an explicit method like `Polynomial.human_readable` that takes some optional parameters (see below).

*3)* Add a standalone function such as `numpy.polynomials.polyutils.format_polynomial` which isn't very OOP but at least it exists and it's out of the way.

*4)* Extend `numpy.set_printoptions` to allow specifying a formatter for Polynomials. This means of course that the actual implementation of the human readable formatting isn't part of numpy but that it is at least easy to add without monkey patching the `Polynomial` class.

**Demo implementation**

I've written a brief sample implementation that can be used to demonstrate the suggested behavior of *1)*. The monkey-patched `Polynomial.__format__` only supports the exponent style option described above, while `format_polynomial()` demonstrates some additional options. You can run the code as is in Python 3. The UTF-8 literals break Python 2 but substituting a compatible implementation should not be an issue.
```Python
exp_styles = {
    'utf8':   lambda n: "".join("⁰¹²³⁴⁵⁶⁷⁸⁹"[int(k)] for k in str(n)),
    'ascii':  "^{:d}".format,
    'latex':  "^{{{:d}}}".format,
    'html':   "<sup>{:d}</sup>".format,
    'python': "**{:d}".format,
}

def format_polynomial(polynomial, variable="x", mul="", exponents=exp_styles['utf8']):
    """
    Format a polynomial given as iterable in a human readable string.

    Parameters:
        polynomial : array_like
        variable : string, optional
            To be used instead of "x" in the formatted output.
        mul : string, optional
            To be used as a multiplication symbol in the output
        exponents : function (str -> str), optional
            Render the exponent from a given integer > 0.

    Examples:
    Format in a Python parseable format:
        >>> s = format_polynomial([-1, 0, 1, 2], "5", "*", "**{}".format)
        >>> print(s)
        2*5**3 + 5**2 - 1
        >>> eval(s)
        274
    """
    if len(polynomial) == 1:
        return format(polynomial[0], "g")
    string = sign = ""
    exps = ("", variable,
            *(variable + exponents(n) for n in range(2, len(polynomial))))
    for n, (coef, exp) in enumerate(zip(polynomial, exps)):  # right to left
        if coef:
            string = exp + sign + string
            if abs(coef) != 1 or n == 0:
                string = format(abs(coef), "g") + (mul if n else "") + string
            sign = " - " if coef < 0 else " + "
    return "-" + string if sign == " - " else string


def _Polynomial__format__(self, style):
    """Monkey patch for Polynomial.__format__"""
    if not style:
        return str(self)
    return format_polynomial(self, exponents=exp_styles[style])

from numpy.polynomial.polynomial import Polynomial
Polynomial.__format__ = _Polynomial__format__


p = Polynomial([-1, 0, 1, 2])
print(format(p))       # default to str(p)
print("{:ascii}".format(p))
# print(f"{p :utf8}")    # Python 3.6+
```

That's it. Please let me know what you think. :+1:

**Comment by charris** [comment]
Might take a look at http://ipython.readthedocs.io/en/stable/api/generated/IPython.core.formatters.html.

When this was originally discussed I didn't want to make rendering the default printing of polynomials, but rather thought that that should be a function of the environment in which Python was being run. If there are some standard hooks, I would favor including them in the base polynomial class. @fperez Thoughts?

**Comment by charris** [comment]
Note also that the various polynomial classes scale and shift the inputs, so that the correct form would be something like:
```
c[0]+ c[1]*(a + b*x) + c[2]*(a + bx)**2 + ...
```

**Comment by philer** [comment]
Isn't numpy independent of IPython? I don't use IPython but maybe it's worth getting into.
In the meantime I'll argue that numpy should be useful on its own and not need to rely on other tools to became fit for human consumption.

I didn't quite understand what that window argument is for until you gave that example. Including that into the formatting won't be difficult although at that point the output may become quite hard to read. I guess it'll be up to the user to decide if he prefers to simplify and convert it into a 'normal' Polynomial.

**Comment by eric-wieser** [comment]
For IPython compatibility, all we need to implement are `_repr_html_` and `_repr_latex_`.

**Comment by charris** [comment]
@philer My argument is that pretty displays are the job of the environment, iPython, Jupyter, etc. If there are a few common standards for such, I'll happily add them.

**Comment by charris** [comment]
Normal polynomials are easy to detect. The variants I would look for in `a + b*x` are `a == 0` and `b == 1`. I'm less clear on what the best way to display that information is. Maybe
```
t = a + b*x
P(t)...
```
For the other polynomial types, need to find some standard polynomial shorthand. Note that ascending powers works better for series than descending powers, so that is another option that need consideration. The precision of the coefficents is another thing to consider, as well as line breaks. Chebyshev series of 100 terms or more are not uncommon.

For classroom use it might be simpler to subclass the polynomial classes and eliminate the bells and whistles, OTOH, it might be better for the students to become familiar with the real thing.

**Comment by mhvk** [comment]
:+1: to the general idea. In `astropy`, we converged on a `to_string()` method that allows one to specify a format, and then using appropriate incantations of those in `__str__`, `__repr__`, `_repr_html`, `_repr_latex`, etc. I've also liked the idea of using `__format__` for this purpose, with the only downside that one already has an existing "mini-language" for formats, and it is not completely obvious the best way forward is to make arbitrary extensions to that.

Overall, that is in fact the most difficult part: shouldn't we have something that is somewhat uniform for all of numpy (e.g., our `Quantity` and `Unit` classes do have latex, html, and unicode representation...).

**Comment by fperez** [comment]
Sympy and Pandas have also adopted the IPython display conventions, to great user benefit. And as @mhvk comments, all that can be done in a way that's not strongly tied to IPython/Jupyter: the main functionality can live in other methods, and the conventions used by IPython are thin shims that call the main/flexible display methods.

That approach gives folks the best of both worlds: easy manual generation of rich representations when desired, and automatically a better UX when working in IPython/Jupyter.

We encourage projects (and try to help with) to offer a good interactive experience in IPython, but obviously we don't want to force/assume anyone to use the tools.

**Comment by eric-wieser** [comment]
Here's what a latex representation could look like:
```python
def PolyBase_repr_latex_(self):
    off, scale = self.mapparms()
    if off == 0 and scale == 1:
        term = 'x'
        parens = False
    elif scale == 1:
        term = '{} + x'.format(off)
        parens = True
    elif off == 0:
        term = '{}x'.format(scale)
        parens = True
    else:
        term = '{} + {}x'.format(off, scale)
        parens = True

    non_zero_coefs = [
        (i, c)
        for i, c in enumerate(self.coef)
        if not (c == 0)  # handle NaN
    ]

    if not non_zero_coefs:
        return '$0$'

    coef_str = [
        '{}'.format(c) if i == 0 else
        ' + {}'.format(c) if c > 0 else
        ' - {}'.format(-c)
        for i, c in non_zero_coefs
    ]
    term_str = [
        self._repr_latex_term(i, term, parens)
        for i, c in non_zero_coefs
    ]

    return '$x \mapsto {}$'.format(''.join(
        '{}\\,{}'.format(c, t) if t != '1' else
        '{}'.format(c)
        for c, t in zip(coef_str, term_str)
    ))



def Polynomial_repr_latex_term(i, term, parens):
    if parens:
        term = r'\left({}\right)'.format(term)
    if i == 0:
        return '1'
    elif i == 1:
        return term
    else:
        return '{}^{{{}}}'.format(term, i)

def Chebyshev_repr_latex_term(i, term, parens):
    return 'T_{{{}}}({})'.format(i, term)

import numpy.polynomial
numpy.polynomial.polynomial.ABCPolyBase._repr_latex_ = PolyBase_repr_latex_
numpy.polynomial.polynomial.Polynomial._repr_latex_term = staticmethod(Polynomial_repr_latex_term)
numpy.polynomial.chebyshev.Chebyshev._repr_latex_term = staticmethod(Chebyshev_repr_latex_term)
```

**Comment by eric-wieser** [comment]
Which gives

![image](https://user-images.githubusercontent.com/425260/42151156-dfeaa150-7d90-11e8-833a-f01b73f4cb6c.png)

**Comment by eric-wieser** [comment]
I've submitted a variant of the above at #11528

**Comment by mattip** [comment]
Does #11528 close this, or is there more to do?
